### PR TITLE
Task/tests

### LINKED ISSUE #48: As a developer, I need a defined test strategy for the CLI
[AB#6765](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/6765)

### PR DESCRIPTION
**Pull request by jobala**
## Overview

Adds **core** and **command module** tests

### Demo

N/A

### Notes

Tests in `msgraph.core.cli` are copied from `azure-cli` and modified to remove `azure-cli` specific implementations. Most of the `azure-cli` tests are left out because they either have a dependency on azure-cli extension support -- which we do not support -- or tests specific `azure-cli` implementations.

## Testing Instructions

* Pull these changes
* Activate a virtual environment `pipenv shell`
* Run `pytest`
* Observe that tests are executed.

Fixes #48, #46

### CODE REVIEW

File: msgraph/cli/core/__init__.py [hunk 1]
```diff
@@ -13,14 +13,24 @@
...
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
+    def refresh_request_id(self):
+        """Assign a new random GUID as x-ms-client-request-id
+        The method must be invoked before each command execution in order to ensure
+        unique client-side request ID is generated.
+        """
+        import uuid
+        self.data['headers'] = {'x-ms-client-request-id': str(uuid.uuid1())}
```
Comment (MIchaelMainer) [inline msgraph/cli/core/__init__.py]: The header name is `client-request-id`. We will want to add a logging handler into Python Core to log this so customers can share this header for support scenarios. I've added #61 to track this.

File: msgraph/cli/core/commands/_util.py [hunk 2]
```diff
@@ -137,3 +140,73 @@ def _merge_kwargs(patch_kwargs, base_kwargs, supported_kwargs=None):
     if unrecognized_kwargs:
         raise TypeError('unrecognized kwargs: {}'.format(unrecognized_kwargs))
     return merged_kwargs
+
+
+def _expand_file_prefixed_files(args):
```
Comment (MIchaelMainer) [inline msgraph/cli/core/commands/_util.py:145]: Please provide comments around what this function and subroutines are intended to do. A couple of examples would be helpful.

File: msgraph/cli/core/commands/_util.py [hunk 3]
```diff
@@ -137,3 +140,73 @@ def _merge_kwargs(patch_kwargs, base_kwargs, supported_kwargs=None):
...
+        arg_split = arg.split('=', 1)
+        try:
+            return '='.join([arg_split[0], _maybe_load_file(arg_split[1])])
+        except IndexError:
+            return _maybe_load_file(arg_split[0])
+
+    return list([_expand_file_prefix(arg) for arg in args])
+
+
+def register_cache_arguments(cli_ctx):
```
Comment (MIchaelMainer) [inline msgraph/cli/core/commands/_util.py]: `register_cache_arguments` doesn't appear to be used. Either remove or add implementation (plus update help docs below).. 

`grep -r "register_cache_arguments" ./`

File: msgraph/cli/core/commands/validators.py [hunk 4]
```diff
@@ -8,6 +11,15 @@ class IterateValue(list):
     pass  # pylint: disable=unnecessary-pass
 
 
+class IterateAction(argparse.Action):  # pylint: disable=too-few-public-methods
```
Comment (MIchaelMainer) [inline msgraph/cli/core/commands/validators.py]: This is imported into msgraph/cli/core/tests/test_application.py but is not used.

File: msgraph/cli/core/file_util.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
+from knack.log import get_logger
+from knack.util import CLIError
+
+from msgraph.cli.core.help._help import CliCommandHelpFile, CliGroupHelpFile
+
+logger = get_logger(__name__)
+
+
+def get_all_help(cli_ctx, skip=True):
```
Comment (MIchaelMainer) [inline msgraph/cli/core/file_util.py]: What is the intention with `get_all_help` as I don't see this used. 

`grep -r "get_all_help" ./`

File: msgraph/cli/core/mock/mock.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+from msgraph.cli.core import MgCLI
+
+
```
Comment (MIchaelMainer) [inline msgraph/cli/core/mock/mock.py]: Let's consider renaming this to something like `MockCLI`. This file would be better placed in *cli/core/test/*.

File: msgraph/cli/core/mock/mock.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        from knack.completion import ARGCOMPLETE_ENV_NAME
+
+        super(DummyCli, self).__init__(cli_name='mg',
+                                       commands_loader_cls=commands_loader_cls
+                                       or MainCommandsLoader,
+                                       help_cls=GraphCliHelp,
+                                       invocation_cls=GraphCliCommandInvoker)
+
+        self.data['headers'] = {
+        }  # the x-ms-client-request-id is generated before a command is to execute
```
Comment (MIchaelMainer) [inline msgraph/cli/core/mock/mock.py]: client-request-id instead of x-ms-client-request-id

File: msgraph/cli/core/__init__.py [hunk 8]
```diff
@@ -13,14 +13,24 @@
...
 from msgraph.cli.core.command_loaders import MainCommandsLoader, ExtensionCommandsLoader
+
 __version__ = '1.0.0'
 
 
 class MgCLI(CLI):
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
+    def refresh_request_id(self):
```
Comment (MIchaelMainer) [inline msgraph/cli/core/__init__.py:25]: We will want to make sure that the Python Core does not add a client-request-id when either a developer or the CLI provides one.

**Review by MIchaelMainer — COMMENTED** [review]
I'll need to review more tomorrow morn.

File: msgraph/cli/core/__init__.py (hunk 1 above)
Comment (jobala) [reply msgraph/cli/core/__init__.py]: For the CLI users can access the request header with the --debug flag.

image

Does this close #61 ?

For the python core, we are tracking this with [#58](https://github.com/microsoftgraph/msgraph-sdk-python-core/issues/58)

File: msgraph/cli/core/__init__.py (hunk 8 above)
Comment (jobala) [reply msgraph/cli/core/__init__.py:25]: Added [#59](https://github.com/microsoftgraph/msgraph-sdk-python-core/issues/59) to track this in Python Core

File: msgraph/cli/core/__init__.py (hunk 1 above)
Comment (MIchaelMainer) [reply msgraph/cli/core/__init__.py]: Yes, we can resolve the the linked item. I assume this was captured before renaming the client-request-id header.

**Review by MIchaelMainer — APPROVED** [review]

File: msgraph/cli/core/__init__.py [hunk 9]
```diff
@@ -13,25 +13,35 @@
...
+        """Assign a new random GUID as client-request-id
+        The method must be invoked before each command execution in order to ensure
+        unique client-side request ID is generated.
+        """
+        import uuid
+        self.data['headers'] = {'client-request-id': str(uuid.uuid1())}
+
     def get_cli_version(self):
+        #TODO: Read version from environment variable
```
Comment (ddyett) [inline msgraph/cli/core/__init__.py:34]: Issue opened for this todo?  There's about 78 todo's in the code so I'm concerned of losing track.

File: msgraph/cli/core/invocation.py [hunk 10]
```diff
@@ -77,11 +78,14 @@ def _is_paged(obj):
 
 class GraphCliCommandInvoker(CommandInvoker):
     def execute(self, args):
+        # TODO: Can't simply be invoked as an event because args are transformed
```
Comment (ddyett) [inline msgraph/cli/core/invocation.py:81]: So is it really a todo or going to stay this way

File: msgraph/cli/core/mglogging.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+from knack.util import ensure_dir
+from azure.core.pipeline.policies._universal import HttpLoggingPolicy
+from msgraph.cli.core.commands.events import EVENT_INVOKER_PRE_CMD_TBL_TRUNCATE
+
+_UNKNOWN_COMMAND = "unknown_command"
+_CMD_LOG_LINE_PREFIX = "CMD-LOG-LINE-BEGIN"
+
+
+class MgCliLogging(CLILogging):
+    _COMMAND_METADATA_LOGGER = 'az_command_data_logger'
```
Comment (ddyett) [inline msgraph/cli/core/mglogging.py]: does it have to be this specific name or can it be called mg

File: msgraph/cli/core/__init__.py (hunk 9 above)
Comment (jobala) [reply msgraph/cli/core/__init__.py:34]: Tracked in #69. All the other TODOs are in code copied over from azure-cli.

File: msgraph/cli/core/invocation.py (hunk 10 above)
Comment (jobala) [reply msgraph/cli/core/invocation.py:81]: This was from code copied over from azure-cli.

File: msgraph/cli/core/mglogging.py (hunk 11 above)
Comment (jobala) [reply msgraph/cli/core/mglogging.py]: Renaming this to mg.

**Pull request merged**